nevergrad/functions/powersystems/core.py

```diff
@@ -0,0 +1,233 @@
+        return np.matmul(output, self.layers[-1])
+
+
+# Real life is more complicated! This is a very simple model.
+class PowerSystem(inst.InstrumentedFunction):
+    """
+    Parameters
+    ----------
+    nint intaum_stocks: number of stocks to be managed
+    depth: number of layers in the neural networks
+    width: number of neurons per hidden layer
+    """
+
+    def __init__(self, num_stocks: int = 13, depth: int = 3, width: int = 3, 
+            year_to_day_ratio: float = 2.,  # Ratio between std of consumption in the year and std of consumption in the day.
+            constant_to_year_ratio: float = 1.,  # Ratio between constant baseline consumption and std of consumption in the year.
+            back_to_normal: float = 0.5,  # Part of the variability which is forgotten at each time step.
+            consumption_noise: float = 0.1,  # Instantaneous variability.
+            num_thermal_plants: int = 7,  # Number of thermal plants.
+            num_years: int = 1,  # Number of years.
+            failure_cost: float = 500.,  # Cost of not satisfying the demand. Equivalent to an expensive infinite capacity thermal plant.
+            ) -> None:
+        # Number of stocks (dams).
+        self.N = num_stocks
+        N = self.N
+        # Parameters describing the problem.
+        self.year_to_day_ratio = year_to_day_ratio
+        self.constant_to_year_ratio = constant_to_year_ratio
+        self.back_to_normal = back_to_normal
+        self.consumption_noise = consumption_noise
+        self.num_thermal_plants = num_thermal_plants
+        self.number_of_years = num_years 
+        self.failure_cost = failure_cost
+        
+        self.average_consumption = self.constant_to_year_ratio * self.year_to_day_ratio
+        self.thermal_power_capacity = self.average_consumption * np.random.rand(self.num_thermal_plants)
+        self.thermal_power_prices = np.random.rand(num_thermal_plants)
+        dam_managers: List[Any] = []
+        for i in range(N):
+            dam_managers += [Agent(10 + N + 2*self.num_thermal_plants, depth, width)]
+        the_dimension = sum([a.GetParamNumbers() for a in dam_managers])
+        self.dam_managers = dam_managers
+        super().__init__(self._simulate_power_system, Instrumentation(inst.var.Array(the_dimension)))
+        self._descriptors.update(num_stocks=num_stocks, depth=depth, width=width)
+
+    def get_num_vars(self) -> List[Any]:
+        return [m.GetParamNumbers() for m in self.dam_managers]
+
+    def _simulate_power_system(self, x: np.ndarray) -> float:
+        failure_cost = self.failure_cost  # Cost of power demand which is not satisfied (equivalent to a expensive infinite thermal group).
+        dam_managers = self.dam_managers
+        for a in dam_managers:
+            assert(len(x) >= a.GetParamNumbers())
+            a.SetParams(np.array(x[:a.GetParamNumbers()]))
+            x = x[a.GetParamNumbers():]
+        assert(len(x) == 0)
```

Comment:
You wouldn't need this assert and the `x = x[a.GetParamNumbers():]` above if you were to assert `assert(len(x) == a.GetParamNumbers())` above

Reply:
No comprendo! x decreases at each pass in the loop.
assert(len(x) == a.GetParamNumbers())  is not true.

Reply:
Oh yes you're right I thought there was something to get all the parameters in one go. I would have done it differently (without modifying x) but I guess this is fine.